Double-check conditional constness in MIR

compiler-errors opened the pull request:
To prevent any unchecked `~const` bounds from leaking through during MIR lowering. 

If this check fails, it will eventually just delay a bug, but for now it reports errors. That error reporting may be redundant if we're calling it from code that already doesn't allow `~const` (i.e. when the `effects` and `const_trait_impl` gates are disabled), but I don't think it's that big of a deal.

edit: This also makes sure that we issue a const stability error if we encounter *any* function with const conditions when `const_trait_impl` is not enabled. This ensures that that feature remains airtight.

compiler-errors committed:
Double-check conditional constness in MIR

To prevent any conditional constness from leaking through during MIR lowering

compiler-errors commented:
r? fee1-dead

fee1-dead commented on tests/ui/traits/const-traits/cross-crate.stocknc.stderr:
```diff
@@ -1,3 +1,9 @@
+error[E0277]: the trait bound `cross_crate::NonConst: ~const cross_crate::MyTrait` is not satisfied
+  --> $DIR/cross-crate.rs:19:5
+   |
+LL |     NonConst.func();
+   |     ^^^^^^^^^^^^^^^
+
```
note that when we constify libcore in the future, this change may be leaking out unstable details (the `~const` syntax) on stable code

bors commented:
:umbrella: The latest upstream changes (presumably #132349) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented on tests/ui/traits/const-traits/cross-crate.stocknc.stderr:
```diff
@@ -1,3 +1,15 @@
+error[E0015]: cannot call non-const fn `<cross_crate::NonConst as cross_crate::MyTrait>::func` in constant functions
```
Duplicated error is NBD imo.

compiler-errors commented on compiler/rustc_const_eval/src/check_consts/check.rs:
```diff
@@ -360,6 +361,68 @@ impl<'mir, 'tcx> Checker<'mir, 'tcx> {
+            } else if tcx.features().const_trait_impl() {
+                infcx.err_ctxt().report_fulfillment_errors(errors);
+            } else {
+                self.check_op(ops::FnCallNonConst {
```
I'm actually leaning towards a different choice here. Instead of selecting these const conditions, we should probably just reject calling *any* functions with nontrivial const conditions if `const_trait_impl` is not enabled.

compiler-errors committed:
Completely deny calling functions with const conditions in MIR const check unless const_trait_impl is enabled

This will help us make sure that we never leak any conditionally const
functions into stable.

fee1-dead reviewed (approved):
lgtm

compiler-errors commented:
Let's test perf

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

fee1-dead commented:
r=me if perf shows nothing significant

bors commented:
:hourglass: Trying commit 314b383678df037ff40c1718b4266413b360babf with merge 7e9717427a68dd27950b5fcc548e08578ab1b2b0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"314b383678df037ff40c1718b4266413b360babf","merge_sha":"7e9717427a68dd27950b5fcc548e08578ab1b2b0"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11615448896/job/32346069424) [(plain)](https://github.com/rust-lang-ci/rust/commit/7e9717427a68dd27950b5fcc548e08578ab1b2b0/checks/32346069424/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
   Compiling rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
error[E0433]: failed to resolve: could not find `HostPolarity` in `ty`
   --> compiler/rustc_const_eval/src/check_consts/check.rs:397:47
    |
397 |             hir::ConstContext::ConstFn => ty::HostPolarity::Maybe,
    |                                               |
    |                                               could not find `HostPolarity` in `ty`
    |                                               help: an enum with a similar name exists: `ImplPolarity`

---
Caused by:
    Command RUST_BACKTRACE=full python3 /checkout/x.py build --target x86_64-unknown-linux-gnu --host x86_64-unknown-linux-gnu --stage 2 library/std --rust-profile-generate /tmp/tmp-multistage/opt-artifacts/rustc-pgo --set llvm.thin-lto=false --set llvm.link-shared=true [at /checkout/obj] has failed with exit code Some(1)

Stack backtrace:
   0: <anyhow::Error>::msg::<alloc::string::String>
             at /rust/deps/anyhow-1.0.89/src/backtrace.rs:27:14
   1: <opt_dist::exec::CmdBuilder>::run
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/exec.rs:80:17
   2: <opt_dist::exec::Bootstrap>::run
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/exec.rs:181:9
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/main.rs:225:13
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/main.rs:225:13
   4: <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}::{closure#0}, ()>
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/timer.rs:111:22
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/main.rs:214:9
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/main.rs:214:9
   6: <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}, opt_dist::training::RustcPGOProfile>
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/timer.rs:111:22
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/main.rs:211:29
   8: opt_dist::main
             at /rustc/7e9717427a68dd27950b5fcc548e08578ab1b2b0/src/tools/opt-dist/src/main.rs:401:18
   9: <fn() -> core::result::Result<(), anyhow::Error> as core::ops::function::FnOnce<()>>::call_once
   9: <fn() -> core::result::Result<(), anyhow::Error> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0125edf41b8ce32db5c2cce0c83e32ad2e4dde17/library/core/src/ops/function.rs:250:5
  10: std::sys::backtrace::__rust_begin_short_backtrace::<fn() -> core::result::Result<(), anyhow::Error>, core::result::Result<(), anyhow::Error>>
             at /rustc/0125edf41b8ce32db5c2cce0c83e32ad2e4dde17/library/std/src/sys/backtrace.rs:154:18
  11: std::rt::lang_start::<core::result::Result<(), anyhow::Error>>::{closure#0}
             at /rustc/0125edf41b8ce32db5c2cce0c83e32ad2e4dde17/library/std/src/rt.rs:195:18
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
  13: std::panicking::try::do_call
             at /rustc/0125edf41b8ce32db5c2cce0c83e32ad2e4dde17/library/std/src/panicking.rs:557:40
  14: std::panicking::try
             at /rustc/0125edf41b8ce32db5c2cce0c83e32ad2e4dde17/library/std/src/panicking.rs:520:19
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11615448896/job/32347079043)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/11615448896/job/32347079043","builder_name":"checks-actions"} -->

compiler-errors commented:
AH, yes, rebase...

compiler-errors commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 505e083517af9b4d2ce6f30315b6ca3fb6c630be with merge 95d3a7902bae7929446953c1dca12f5e2465aa63...
<!-- homu: {"type":"TryBuildStarted","head_sha":"505e083517af9b4d2ce6f30315b6ca3fb6c630be","merge_sha":"95d3a7902bae7929446953c1dca12f5e2465aa63"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11617042692/job/32356348309)
Build commit: 95d3a7902bae7929446953c1dca12f5e2465aa63 (`95d3a7902bae7929446953c1dca12f5e2465aa63`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11617042692/job/32356348309"},"merge_sha":"95d3a7902bae7929446953c1dca12f5e2465aa63"} -->

rust-timer commented:
Queued 95d3a7902bae7929446953c1dca12f5e2465aa63 with parent 20c909ff9cdd88d33768a4ddb8952927a675b0ad, future [comparison URL](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=95d3a7902bae7929446953c1dca12f5e2465aa63).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (95d3a7902bae7929446953c1dca12f5e2465aa63): [comparison URL](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=95d3a7902bae7929446953c1dca12f5e2465aa63&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=95d3a7902bae7929446953c1dca12f5e2465aa63&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.0% | [0.8%, 1.3%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=95d3a7902bae7929446953c1dca12f5e2465aa63&stat=max-rss)
<details>
<summary>Results (secondary 2.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.2%, 2.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=95d3a7902bae7929446953c1dca12f5e2465aa63&stat=cycles:u)
<details>
<summary>Results (secondary -5.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.3% | [-7.1%, -3.5%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=20c909ff9cdd88d33768a4ddb8952927a675b0ad&end=95d3a7902bae7929446953c1dca12f5e2465aa63&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 784.059s -> 784.63s (0.07%)
**Artifact size**: 333.59 MiB -> 333.57 MiB (-0.01%)

compiler-errors commented:
@bors r=fee1-dead

bors commented:
:pushpin: Commit 505e083517af9b4d2ce6f30315b6ca3fb6c630be has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead 505e083517af9b4d2ce6f30315b6ca3fb6c630be -->
<!-- homu: {"type":"Approved","sha":"505e083517af9b4d2ce6f30315b6ca3fb6c630be","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #132435) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented:
@bors r=fee1-dead

bors commented:
:hourglass: Testing commit 57f2e12f4aafb2c37f45090be97c68c5020c938f with merge 705cfe0e966399e061d64dd3661bfbc57553ed87...
<!-- homu: {"type":"BuildStarted","head_sha":"57f2e12f4aafb2c37f45090be97c68c5020c938f","merge_sha":"705cfe0e966399e061d64dd3661bfbc57553ed87"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11634913305/job/32408889117)
Approved by: fee1-dead
Pushing 705cfe0e966399e061d64dd3661bfbc57553ed87 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"fee1-dead","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11634913305/job/32408889117"},"merge_sha":"705cfe0e966399e061d64dd3661bfbc57553ed87"} -->

rust-timer commented:
Finished benchmarking commit (705cfe0e966399e061d64dd3661bfbc57553ed87): [comparison URL](https://perf.rust-lang.org/compare.html?start=5ca0e9fa9b2f92b463a0a2b0b34315e09c0b7236&end=705cfe0e966399e061d64dd3661bfbc57553ed87&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=5ca0e9fa9b2f92b463a0a2b0b34315e09c0b7236&end=705cfe0e966399e061d64dd3661bfbc57553ed87&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5ca0e9fa9b2f92b463a0a2b0b34315e09c0b7236&end=705cfe0e966399e061d64dd3661bfbc57553ed87&stat=max-rss)
<details>
<summary>Results (secondary 1.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.1%  | [2.3%, 4.0%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.3% | [-3.3%, -3.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=5ca0e9fa9b2f92b463a0a2b0b34315e09c0b7236&end=705cfe0e966399e061d64dd3661bfbc57553ed87&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=5ca0e9fa9b2f92b463a0a2b0b34315e09c0b7236&end=705cfe0e966399e061d64dd3661bfbc57553ed87&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 780.79s -> 780.514s (-0.04%)
**Artifact size**: 334.96 MiB -> 334.93 MiB (-0.01%)

RalfJung commented on compiler/rustc_const_eval/src/check_consts/check.rs:
```diff
@@ -360,6 +361,73 @@ impl<'mir, 'tcx> Checker<'mir, 'tcx> {
+        // If there are any const conditions on this fn and `const_trait_impl`
+        // is not enabled, simply bail. We shouldn't be able to call conditionally
+        // const functions on stable.
+        if !const_conditions.is_empty() && !tcx.features().const_trait_impl() {
```
You should never call `tcx.features()` in this code, as that won't properly check for `rustc_allow_const_fn_unstable`.

compiler-errors replied:
Well, the user still needs to enable the feature gate *globally* so that we can properly enforce the feature's validity earlier in the compiler.

RalfJung commented on compiler/rustc_const_eval/src/check_consts/check.rs:
```diff
@@ -584,31 +651,7 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
-                if !errors.is_empty() {
-                    infcx.err_ctxt().report_fulfillment_errors(errors);
-                }
+                self.revalidate_conditional_constness(callee, fn_args, call_source, *fn_span);
```
I am confused, why is this called outside the `if let Some(trait_did) = tcx.trait_of_item(callee) {`? Can there be non-trait conditional const calls? Given that the `const_trait_impl` feature gate is used for these calls, I assume the answer is no...

compiler-errors replied:
Yes? We can put ~const bounds on free const functions.

compiler-errors replied:
>  Given that the const_trait_impl feature gate is used for these calls

I think you're getting too stuck on the naming of the `const_trait_impl` feature gate. 

The feature gate enables the declaration of `const` traits, `impl const`, and the use of `~const` trait bounds on conditionally const items, which includes free const function. This is how the gate has been used for the last like... 3 years or whatever.

I don't want to rename the feature gate, since that basically. the feature gate could be called `foo_bar` for all it matters.

RalfJung replied:
That's a poorly chosen name. :(

compiler-errors replied:
🤷 I'm sorry, but I didn't choose it, and I'm not really interested in doing a rename here. That can be done later, and ideally not when there are like 3 other PRs in flight concerning it.
